Client/Assets/ContextMenu.js

```diff
@@ -97,50 +111,52 @@ addEventListener("touchstart", function (event) {
     }
 
     var data = {};
-    var element = event.target;
-    var style = getComputedStyle(element);
 
     // Don't show the context menu if this element is has the
     // -webkit-touch-callout: none CSS property applied.
+    var style = getComputedStyle(event.target);
     if (style.webkitTouchCallout === "none") {
       cancel();
       return;
     }
 
-    // Listen for touchend or move events to cancel the context menu timeout.
-    element.addEventListener("touchend", handleTouchEnd);
-    element.addEventListener("touchmove", handleTouchMove);
-
-    do {
-      if (!data.link && element.localName === "a") {
-        data.link = element.href;
-
-        // The web view still shows the tap highlight after clicking an element,
-        // so add a delay before showing the long press highlight to avoid
-        // the highlight flashing twice.
-        var linkElement = element;
-        setTimeout(function () {
-          if (longPressTimeout) {
-            createHighlightOverlay(linkElement);
-          }
-        }, 100);
-      }
-
-      if (!data.image && element.localName === "img") {
-        data.image = element.src;
-      }
+    element = event.target.closest("a");
```

Comment:
Instead of a `do`/`while` loop, we can use the "new" `closest()` DOM API which is available as of iOS 9. This method will search up the tree from the given element for the specified CSS selector. If the original target element *matches* the CSS selector itself, then it simply returns the original element and stops traversal of the tree immediately. Our previous code was essentially an implementation of this exact behavior, but it was very inefficient because it continued traversing the tree all the way to the root element even after it had found a match.

Reply:
I should also note that the reason this is needed in the first place is because `event.target` may not itself be an `<a href="..."></a>` element when clicking a link. As an example:

```
<a href="/some/link">
  <div>
    <span>Click me!</span>
  </div>
</a>
```

With the above HTML, clicking on the link will likely send the `<span>` element as the `event.target`. So, we always need to traverse from `event.target` up the tree until we find an `<a>...</a>` (if one exists).